crudd/src/main.rs

```diff
@@ -0,0 +1,510 @@
+    } else {
+        // Account for the bytes we already read, and add one to the block
+        // offset
+        (
+            num_bytes - alignment_bytes,
+            opt.byte_offset / native_block_size + 1,
+        )
+    };
+
+    // How many full commands we can issue, and then what our remainder bytes
+    // will be. Note we'll round up the remainder later to get the number of
+    // blocks for the last iocmd
+    let cmd_count = num_bytes / (opt.iocmd_block_count * native_block_size);
+    let remainder = num_bytes % (opt.iocmd_block_count * native_block_size);
+
+    // We need to run this at the end to handle the remainder bytes, but we also
+    // need to do the exact same thing if the input stream ends early. So
+    // let's not copy-paste
+    // - write all block-aligned data remaining
+    // - read/mod/write the last block if necessary
+    // - issue a flush
+    // - block on all waiters
+    let write_remainder_and_finalize = |mut w_buf: BytesMut,
+                                        offset: Block,
+                                        n_read: usize,
+                                        mut waiters: VecDeque<
+        BlockReqWaiter,
+    >|
+     -> Result<()> {
+        // the input stream ended,
+        // - read/mod/write for alignment
+        // - block on all waiters
+
+        // uflow short for underflow, as we're underflowing are normal iocmd
+        // block
+        let uflow_blocks = n_read as u64 / native_block_size;
+        let uflow_remainder = n_read as u64 % native_block_size;
+
+        if uflow_remainder == 0 {
+            // no need to RMW, just write
+            w_buf.resize(n_read, 0);
+            let w_waiter = crucible.write(offset, w_buf.freeze())?;
+            waiters.push_back(w_waiter);
+        } else {
+            // RMW oof
+
+            // Adjust buffer to be block-aligned to our data
+            w_buf.resize(((uflow_blocks + 1) * native_block_size) as usize, 0);
+
+            // How many bytes we need to copy into the end of w_buf
+            let uflow_backfill = (native_block_size - uflow_remainder) as usize;
+
+            // First, read the final partial-block
+            let uflow_offset = Block::new(
+                offset.value + uflow_blocks,
+                native_block_size.trailing_zeros(),
+            );
+            let uflow_r_buf = Buffer::new(native_block_size as usize);
+            let mut r_waiter =
+                crucible.read(uflow_offset, uflow_r_buf.clone())?;
+            r_waiter.block_wait()?;
+
+            // Copy it into w_buf
+            let r_bytes = uflow_r_buf.as_vec();
+            w_buf[n_read..n_read + uflow_backfill]
+                .copy_from_slice(&r_bytes[uflow_remainder as usize..]);
+
+            // Issue the write
+            let w_waiter = crucible.write(offset, w_buf.freeze())?;
+            waiters.push_back(w_waiter);
+        }
+
+        // Flush
+        let flush_waiter = crucible.flush(None)?;
```

Comment:
I don't think you technically need a flush here.  It's nice to do if you like your data, but I also wonder about any impact it has on performance.  As long as you don't have any overlapping IOs, everything should be written out as you expect.